tests/create_test_disks.sh

```diff
@@ -0,0 +1,90 @@
+#!/bin/bash
+
+set -e
+
+LOOP_DEVICE_MAJOR=7
+LOOP_DEVICE_MIN_ID=100
+
+echo_err() {
+    echo "$@" >&2
+}
+
+create_device_node() {
+    local -r last_id=$(find /dev/ -name 'loop[0-9]*' | sed 's/^.*loop//' | sort -r -n | head -1)
+    local id=$((last_id + 1))
+    if [[ "$id" -lt "$LOOP_DEVICE_MIN_ID" ]]; then
+        id="$LOOP_DEVICE_MIN_ID"
+    fi
+    local path
+    for (( i=0; i<10; i++ )); do
+        path="/dev/loop$id"
+        if [[ ! -e "$path" ]] && sudo mknod "$path" b "$LOOP_DEVICE_MAJOR" "$id"; then
+            echo "$path"
+            return 0
+        fi
+        $((id++))
+    done
+    return 1
+}
+
+device_mountpoint() {
+    local -r datadir="$1"
+    local -r dev="$2"
+    if [[ -z "$datadir" || -z "$dev" ]]; then
+        echo_err "Usage: device_mountpoint <data dir> <device>"
+        return 1
+    fi
+    echo "$datadir/mnt-$(basename "$dev")"
+}
+
+create_device() {
+    local -r datadir="$1"
+    if [[ -z "$datadir" ]]; then
+        echo_err "Usage: create_device <data dir>"
+        return 1
+    fi
+    local -r dev=$(create_device_node)
+    local -r fs="$datadir/$(basename "$dev").vhd"
+    local -r mountpoint=$(device_mountpoint "$datadir" "$dev")
+    echo_err
+    echo_err "# Device $dev"
+    dd if=/dev/zero of="$fs" bs=64K count=128 >/dev/null 2>&1
+    sudo losetup "$dev" "$fs"
```

Comment:
some environments might not have `sudo` installed or prefer another tool to elevate privileges (e.g. `doas`).

I think it would be better to remove the use of `sudo` in the script and add something at the beginning like:

```
if [[ $EUID -ne 0 ]]; then
   echo "This script must be run as root." >&2
   exit 1
fi
```

Reply:
Requiring root for the whole script obfuscates which operations require special privileges and makes the functions within the script less helpful (the current structure permits defining these functions in an interactive shell and then calling them as desired; wrapping the invocation in sudo or doas loses the shell customizations, including the function definitions). I will just make sudo optional.